File: hahlw/array/55.跳跃游戏.cpp
/*
 * @lc app=leetcode.cn id=55 lang=cpp
 *
 * [55] 跳跃游戏
 */
class Solution {
public:
    /*
    原则：如果当前的跳跃范围超过了n-1，则为true。
    不断更新跳跃范围。
    */
    bool canJump(vector<int>& nums) {
        int cur_max=nums[0]+0,n=nums.size(),j=0;
        if(n==0||n==1)return true;
        while(j<=cur_max){
            if(cur_max>=n-1)return true;
            cur_max=max(cur_max,nums[j]+j);
            j++;
        }
        return false;
    }
};

File: Juniper/55.跳跃游戏.py
#
# @lc app=leetcode.cn id=55 lang=python3
#
# [55] 跳跃游戏
#
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        lens = len(nums)
        if lens == 1:
            return True
        elif not nums[0]:
            return False
        def find(num, length):
            for i in range(length-1):
                if num[i] >= length-1 - i :
                    while num[i-1] >=1 and i > 0:
                        i = i-1
                    if i ==0:
                        return True
                    return True and find(num[:i+1], i+1)
            return False
        return find(nums,lens)

        

